src/bench/logging.cpp

```diff
@@ -17,32 +17,61 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
     bench.run([&] { log(); });
 }
 
-static void LoggingYoThreadNames(benchmark::Bench& bench)
+static void LogPrintLevelWithThreadNames(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=1"}, [] { LogPrintf("%s\n", "test"); });
+    Logging(bench, {"-logthreadnames=1"}, [] { LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Error, "%s\n", "test"); });
 }
-static void LoggingNoThreadNames(benchmark::Bench& bench)
+
+static void LogPrintLevelWithoutThreadNames(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=0"}, [] { LogPrintf("%s\n", "test"); });
+    Logging(bench, {"-logthreadnames=0"}, [] { LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Error, "%s\n", "test"); });
+}
+
+static void LogPrintWithCategory(benchmark::Bench& bench)
+{
+    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```

Comment:
What's the expected behavior when logging with a category that is not specified in `-debug=`?

Does logging with `BCLog::VALIDATION` but with `-debug=net` means that `BCLog::Logger::WillLogCategoryLevel` will return `false`? Therefore `LogPrint()` will not actually log?

Reply:
> What's the expected behavior when logging with a category that is not specified in -debug=?

It should not log. Note that the `LogPrintWithoutCategory` benchmark specifies `-debug=0` which disables all categories, although this has a problem that I mentioned in my other comment.

> Does logging with BCLog::VALIDATION but with -debug=net means that BCLog::Logger::WillLogCategoryLevel will return false? Therefore LogPrint() will not actually log?

That's correct.

Reply:
This is different from what I observe running this benchmark.

When `BCLog::NET` was replaced with `BCLog::VALIDATION` it seems that changing `-debug=net` to `-debug=validation` was overlooked. Which per my understanding, means that `LogPrintWithCategory` should not actually benchmark what is supposed to, but rather be comparable to `LogPrintWithoutCategory`. This is not true and I only see the expected behavior when I change `-debug=net` to `-debugexclude=validation`.

I will investigate this further.

Reply:
I think I figured out this mystery. It's a wild story. I'm almost certain you're right that the test has a mistake in it:
```
Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
This is meant to specify the _same_ category (not the different categories, "net" and `VALIDATION`). But the confusion is that with this mistake, the benchmark still shows a large `ns/op`, as if it's actually logging. I ran this code with the debugger, which shows that it is indeed logging the message. The reason, I just figured out, is that the test framework [enables _all_ logging](https://github.com/bitcoin/bitcoin/blob/7241b936c5b4d004cdbdc7dc4d0b586856f6716b/src/test/util/setup_common.cpp#L113), so when the test specifies `-debug=net`, that has no effect because all categories are enabled anyway. 

I tried changing that line of benchmark code to:
```
Logging(bench, {"-logthreadnames=0", "-debug=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
Note the addition of `-debug=0` which should disable all logging. So now what should happen is, the test framework specifies `-debug` which enables all categories (the test framework arguments occur before the ones we specify here), but then we specify `-debug=0` which _disables_ all categories, then add back just the one we want. Now, no logging should actually happen, since we haven't enabled `"validation"`. And this was the case, no logging occurred (the `ns/op` was a low number).

Now, let's just fix the bug in the test:
```
Logging(bench, {"-logthreadnames=0", "-debug=0", "-debug=validation"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
To my surprise, this also had a low `ns/op`, meaning that it was not logging! How could that be, when we're explicitly specifying `-debug=validation` and we're doing a `VALIDATION` print?

Further investigation revealed that when the command line includes `-debug=0`, then we [_do not_ enable](https://github.com/bitcoin/bitcoin/blob/7241b936c5b4d004cdbdc7dc4d0b586856f6716b/src/init/common.cpp#L86) _any categories_ specified on the command line! That is, the `-debug=0` causes even later `-debug=category` arguments to be ignored!

Anyway, I believe this is the fix we want, this gives the expected results (actual logging, a high `ns/op`):
```diff
 static void LogPrintWithCategory(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
+     // the test framework enables all categories by default
+    Logging(bench, {"-logthreadnames=0"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
 }
```